indico/modules/events/surveys/controllers/display.py

```diff
@@ -110,15 +122,39 @@ def _process(self):
             back_button_endpoint = None
         return self.view_class.render_template('display/survey_questionnaire.html', self._conf, form=form,
                                                event=self.event_new, survey=self.survey,
-                                               back_button_endpoint=back_button_endpoint)
+                                               back_button_endpoint=back_button_endpoint,
+                                               partial_completion=self.survey.partial_completion)
+
+    def _make_form(self):
+        survey_form_class = make_survey_form(self.survey)
+        if self.submission and request.method == 'GET':
+            return survey_form_class(formdata=MultiDict(self.submission.pending_answers))
+        else:
+            return survey_form_class()
 
-    def _save_answers(self, form):
+    @no_autoflush
+    def _save_answers(self, form, submission):
         survey = self.survey
-        submission = SurveySubmission(survey=survey)
-        if not survey.anonymous:
-            submission.user = session.user
+        if not submission:
+            submission = SurveySubmission(survey=survey, user=session.user)
+        submission.is_anonymous = survey.anonymous
         for question in survey.questions:
             answer = SurveyAnswer(question=question, data=getattr(form, 'question_{}'.format(question.id)).data)
             submission.answers.append(answer)
         db.session.flush()
         return submission
+
+
+class RHSaveSurveyAnswers(RHSubmitSurveyBase):
+    def _checkProtection(self):
+        RHSubmitSurveyBase._checkProtection(self)
+        if not self.survey.partial_completion:
+            raise Forbidden(response=redirect_to_login(reason=_('You are trying to save a survey '
```

Comment:
Why redirect to login? I'd just `raise Forbidden` - no need to specify messages when it's something only reachable by tampering with the data sent